hotshot-state-prover/src/bin/gen-demo-contract.rs

```diff
@@ -0,0 +1,157 @@
+}
+
+fn parse_seed(s: &str) -> Result<[u8; 32], ParseSeedError> {
+    <[u8; 32]>::try_from(hex::decode(s)?)
+        .map_err(|vec| ParseSeedError::WrongLength { length: vec.len() })
+}
+
+pub fn stake_table_commitment_for_demo(
+    num_nodes: usize,
+    keygen_seed: [u8; 32],
+) -> ((CircuitField, CircuitField, CircuitField), U256) {
+    // We now initialize a static stake table as what hotshot orchestrator does.
+    // In the future we should get the stake table from the contract.
+    let mut st = StakeTable::<BLSPubKey, StateVerKey, CircuitField>::new(STAKE_TABLE_CAPACITY);
+    (0..num_nodes).for_each(|id| {
+        let bls_key = BLSPubKey::generated_from_seed_indexed(keygen_seed, id as u64).0;
+        let state_ver_key =
+            StateKeyPair::generate_from_seed_indexed(keygen_seed, id as u64).ver_key();
+        st.register(bls_key, U256::from(1u64), state_ver_key)
+            .expect("Key registration shouldn't fail.");
+    });
+    st.advance();
+    st.advance();
+    (
+        st.commitment(SnapshotVersion::LastEpochStart)
+            .expect("Commitment computation shouldn't fail."),
+        st.total_stake(SnapshotVersion::LastEpochStart).unwrap() * 2 / 3,
+    )
+}
+
+fn main() {
+    let args = Args::parse();
+
+    let ((bls_comm, schnorr_comm, stake_comm), threshold) =
+        stake_table_commitment_for_demo(args.num_nodes, args.keygen_seed);
+
+    // calculate the path to solidity file
+    let contract_name = "LightClient.s.sol";
+    let mut path = args.contracts_root_dir;
+    path.push("script");
+    path.push(contract_name);
+    path.set_extension("sol");
+
+    println!(
+        "Writing contract into file: {}",
+        path.as_os_str().to_str().unwrap()
+    );
+    // overwrite the file
+    let mut file = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .open(path)
+        .unwrap();
+
+    let code = format!(
+        "// SPDX-License-Identifier: UNLICENSED
+        pragma solidity ^0.8.0;
+        
+        import \"forge-std/Script.sol\";
+        import {{ BN254 }} from \"bn254/BN254.sol\";
+        import {{ LightClient as LC }} from \"../src/LightClient.sol\";
+        
+        contract DeployLightClientScript is Script {{
+            function run() external {{
+                string memory seedPhrase = vm.envString(\"MNEMONIC\");
+                uint256 privateKey = vm.deriveKey(seedPhrase, 0);
+                vm.startBroadcast(privateKey);
+        
+                // For Decaf there will be only one epoch
+                uint32 blocksPerEpoch = type(uint32).max;
+        
+                uint64 viewNum = 0;
+                uint64 blockHeight = 0;
+                BN254.ScalarField blockCommRoot = BN254.ScalarField.wrap(0);
+                BN254.ScalarField feeLedgerComm = BN254.ScalarField.wrap(0);
+                BN254.ScalarField stakeTableBlsKeyComm = BN254.ScalarField.wrap(
+                    {}
```

Comment:
I don't quite understand why we have to generate this code. Couldn't we use env vars to pass arguments to the deploy script instead?

Reply:
>Couldn't we use env vars to pass arguments to the deploy script instead?

that's a good point, maybe we can use [`vm.envString()` ](https://book.getfoundry.sh/cheatcodes/env-string?#envstring) to load instead. 

i guess the complication comes from: the desire to do "one-click" demo deployment, whereas the mock genesis state requires running rust to determine. 
^^ can we deterministically compute this and hardcode them into `.env`, maybe?

Reply:
the initial motivation is: we could have some different genesis state (different num_nodes, diff stake table config), which would result in variable stake table commitments, 
thus, Chengyu wrote this generation file to automate/simplify the hardcoding process.

Reply:
I guess this answers my question above regarding how commitment values are generated.

Reply:
you already have some deployment script in https://github.com/EspressoSystems/espresso-sequencer/pull/1079,

may I suggest:
- relocate this code to `contracts/rust/gen-light-client-deploy` (basically at the same level as `gen-vk-contract` binary), as it's harder to find this code that supposed to generate script in `contracts/scripts` 
- in your 1079, Philippe you try to adapt this generated by rust approach instead of `diff-test mock-gensis`?

cc @philippecamacho @mrain

Reply:
@sveitser's approach looks more straightforward and flexible (we could have different tests with different initial states for the light client contract) to me.

Reply:
The problem of environment variable way is, computation of the genesis state is not that straightforward.

Reply:
Why not using a forge deploy script and generate the genesis state using `diff-test mock-genesis`? That way in the test you can generate the state and then call the deploy script. See example here: https://github.com/EspressoSystems/espresso-sequencer/blob/0e4a83e4486de1580711643d53eb8cc236d07aff/contracts/test/LightClientContractUpgrade.t.sol#L26 

In this example we generate the genesis state inside the [`deployer.run()`](https://github.com/EspressoSystems/espresso-sequencer/blob/0e4a83e4486de1580711643d53eb8cc236d07aff/contracts/script/DeployLightContract.s.sol#L14) function but that could be done outside.

Reply:
As agreed with @mrain , we will revisit how the light contract is initialized in PR #1079.